tests/test_dydx.py

```diff
@@ -0,0 +1,196 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import json
+import os
+import time
+
+
+from pymaker import Wad
+from pyexchange.dydx import DydxApi, Order, Trade
+
+# Even though DyDx is a Decentralized Exchange,
+# supporting infrastructure required for the orderbook is only available on mainnet.
+# Therefore, mock data was used.
+
+# Models HTTP response, produced by DydxMockServer
+class MockedResponse:
+    def __init__(self, text: str, status_code=200):
+        assert (isinstance(text, str))
+        assert (isinstance(status_code, int))
+        self.status_code = status_code
+        self.ok = 200 <= status_code < 400
+        self.text = text
+        self.reason = None
+
+    def json(self, **kwargs):
+        return json.loads(self.text)
+
+# Determines response based upon the DyDx Client method used
+class DydxMockServer:
+    # Read JSON responses from a pipe-delimited file, avoiding JSON-inside-JSON parsing complexities
+    responses = {}
+    cwd = os.path.dirname(os.path.realpath(__file__))
+    response_file_path = os.path.join(cwd, "mock/dydx-api-responses")
+    with open(response_file_path, 'r') as file:
+        for line in file:
+            kvp = line.split("|")
+            assert(len(kvp) == 2)
+            responses[kvp[0]] = kvp[1]
+
+    @staticmethod
+    def handle_get_pairs(**kwargs):
+        return MockedResponse(text=DydxMockServer.responses["markets"]).json()
+
+    @staticmethod
+    def handle_get_balances(**kwargs):
+        return MockedResponse(text=DydxMockServer.responses["balances"]).json()
+
+    @staticmethod
+    def handle_get_orders(**kwargs):
+        return MockedResponse(text=DydxMockServer.responses["orders"]).json()
+
+    @staticmethod
+    def handle_place_order(**kwargs):
+        return MockedResponse(text=DydxMockServer.responses["place_order"]).json()
+
+    @staticmethod
+    def handle_cancel_order(**kwargs):
+        return MockedResponse(text=DydxMockServer.responses["cancel_order"]).json()
+
+    @staticmethod
+    def handle_get_trades(**kwargs):
+        return MockedResponse(text=DydxMockServer.responses["trades"]).json()
+
+class TestDydx:
+    def setup_method(self):
+        self.dydx = DydxApi(
+            "https://mainnet.infura.io/v3/cde88afeb2d9475aa8a7b3813e7386cf",
+            "dcba44978751342a68e81b0e487de87e52720f6f94792cc237045bce0f9d05fc"
```

Comment:
is this a fake private key or the real deal?

Reply:
real private key, but for an unused account